keps/sig-architecture/20190412-conformance-behaviors.md

```diff
@@ -0,0 +1,342 @@
+
+* Creation and read of the resource with only required fields specified.
+  * API functions as expected: Resource is created and may be read, and defaults
+    are set. This is mechanical and can be completely generated.
+  * Cluster behaves as expected. This cannot be generated, but a skeleton can be
+    generated that allows test authors to evaluate the condition of the cluster
+    to make sure it meets the expectations.
+* Deletion of the resource. This may be mostly mechanical but if there are side-
+  effects, such as garbage collection of related resources, we may want to have
+  manually written evaluation here as well.
+* Creation of resource with each field set, and update of each mutable field.
+  * For each mutable field, apply a patch to the based resource definition
+    before creation (for create tests), or after creation (for update tests).
+  * Evaluate that the API functions as expected; this is mechanical and
+    generated.
+  * Evaluate that the cluster behaves as expected. In some cases this may be
+    able to re-use the same evaluation function used during the creation tests,
+    but often it will require hand-crafted code to test the conditions.
+    Nonetheless, the scaffolding can be generated, minimizing the effort needed
+    to implement the test.
+
+Additional, handwritten tests will be needed that modify the resource in
+multiple ways and evaulate the behavior. The scaffolding must be built such that
+the same process is used for these tests. The test author must only need to
+define:
+* A patch to apply to create or update the resource.
+* A function to evaluate the effect of the API call.
+
+With those two, the same creation and update scaffolding defined for individual
+field updates can be reused.
+
+### Coverage Tooling
+
+In order to tie behaviors back to the tests that are generated, including
+existing e2e tests that already cover behaviors, new tags with behavior IDs will
+be added to conformance tests. Using the existing conformance framework
+mechanism allows incremental adoption of this proposal. Thus, rather than a new
+conformance framework function, test authors will indicate the behaviors covered
+by their tests with a tag in the `framework.ConformanceIt` call.
+
+This also enables a single test to validate multiple behaviors, although that
+should be discouraged.
+
+### Risks and Mitigations
+
+The behavior definitions may not be properly updated if a change is made to a
+feature, since these changes are made in very different areas in the code.
+However, given that the behaviors defining conformance are generally stable,
+this is not a high risk.
+
+## Design Details
+
+## Implementation History
+
+- 2019-04-12: Created
+- 2019-06-11: Updated to include behavior and test generating from APIs.
+
+## Drawbacks
+
+* Separating behaviors into a file that is not directly part of the test suite
+  creates an additional step for developers and could lead to divergence.
```

Comment:
Since the conformance-wg exists I guess I am not too worried about this. If it was just left up to individual dev self-control I would, but these things would only have to be monitored closely when tests become conformance or not conformance behaviors decided upon, right? In which case I think that is a low risk that we'll suddenly forget about it.